Variable TCP Telnet ports via web interface

fuzzball03 opened the pull request:
Per issue #167 ... Here ya go, first attempt, untested.
Spent an hour or so at work on this today...

I've only done really simple C coding, this will be my first that consists of multiple headers.
Please be easy on me - CONSTRUCTIVE CRITICISM WELCOME.

Note - this is untested.

fuzzball03 committed:
Update home.html

Updates for changing/selecting telnet ports

fuzzball03 committed:
Update ui.js

Updates for changing/selecting telnet ports

fuzzball03 committed:
Update cgipins.c

Fixed formatting/extra space

fuzzball03 committed:
Update config.h

Updates for changing/selecting telnet ports.
Fixed formatting.

fuzzball03 committed:
Update config.h

Moved new variable at bottom as requested in this header comments.

fuzzball03 committed:
Update config.c

Updates for changing/selecting telnet ports

fuzzball03 committed:
Create cgitelnet.h

Updates for changing/selecting telnet ports

fuzzball03 committed:
Update cgitelnet.h

fuzzball03 committed:
Create cgitelnet.c

Todo: Missing function for init in serbridge?

fuzzball03 committed:
Update serbridge.c

Updates for changing/selecting telnet ports
-Add function to reinit serbridge after change Telnet ports.

fuzzball03 committed:
Update serbridge.c

fuzzball03 committed:
Update main.c

Updates for changing/selecting telnet ports

fuzzball03 committed:
Update main.c

fuzzball03 committed:
Update main.c

fuzzball03 committed:
Update main.c

Make sure we use our custom telnet ports at bootup

fuzzball03 committed:
Update cgitelnet.c

Fixed bad variable reference

fuzzball03 committed:
Update cgitelnet.c

Fix implicit function call.

fuzzball03 committed:
Update serbridge.c

Remove unecssary function

fuzzball03 committed:
Update cgitelnet.c

Fixed a stupid typo in a function name.

fuzzball03 committed:
Update cgitelnet.c

Changed function for unsigned integers.

fuzzball03 committed:
Update config.h

Changed flashconfig.telnet_port to unsigned 16bit integer.

fuzzball03 committed:
Update cgitelnet.c

Changed functions for unsigned 16bit integer.

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  int8_t ok = 0;
+  uint8_t port1, port2;
```
ports are 16 bits: 0..65535

fuzzball03 replied:
Of course! I got the thought of a short int = 8bit stuck in my head yesterday.
Once I realized that... unsigned 16bit should get me there.

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,79 @@
+  uint8_t port1, port2;
+  ok |= getUInt16Arg(connData, "port1", &port1);
+  ok |= getUInt16Arg(connData, "port2", &port2);
+  if (ok < 0) return HTTPD_CGI_DONE;
```
You can't just return here, you have to send some HTTP response back. A 304 not modified or a 200 OK would be reasonable choices.

fuzzball03 replied:
Ill modify it to include the responses.

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,79 @@
+  }
+  return HTTPD_CGI_DONE;
+
+collision: {
```
Please put this inline since there is only one "goto" to here.

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,79 @@
+
+    // apply the changes
+    //serbridgeReinit();
+    serbridgeInit(flashConfig.telnet_port1, flashConfig.telnet_port2);
```
Please apply change after saving otherwise you might apply the changes, fail to save and return an error, but then things are inconsistent.

fuzzball03 replied:
Good point

tve commented on esp-link/cgitelnet.h:
```diff
@@ -0,0 +1,9 @@
+#include "httpd.h"
+
+int cgiTelnet(HttpdConnData *connData);
+// int8_t telnet_port1, telnet_port2;
```
you can delete this comment...

fuzzball03 replied:
Will do.

tve commented on esp-link/main.c:
```diff
@@ -96,6 +97,8 @@ HttpdBuiltInUrl builtInUrls[] = {
   { "/services/update", cgiServicesSet, NULL },
   { "/pins", cgiPins, NULL },
+  { "/telnet/info", cgiTelnet, NULL},
+  { "/telnet/update", cgiTelnet, NULL},
```
You only need one since you distinguish using GET and POST. Not sure why there are two for services, something for me to check.

fuzzball03 replied:
Yea, I picked up on this. However I wanted to attempt to avoid someone making an easy mistake of sending a PUT request via '/telnet' & accidently modifying it.
However, ill be happy to merge it to just one.
Also, ill have to modify the Ajax method for making inputs since it makes a call to /update

Services uses two because this & pins cgi use a single method to call back set or get. They determine which to use based on the header received(or GET or PUT)

tve commented on serial/serbridge.c:
```diff
@@ -469,6 +469,8 @@ serbridgeInitPins()
 
 // Start transparent serial bridge TCP server on specified port (typ. 23)
+// Here is where we need to change the ports. But how do we do this from the Ajax call?
+
```
Delete left-over comment? You have that here https://github.com/jeelabs/esp-link/pull/205/files#diff-b6cc56038011d1226a2ee85cb280b897R47

fuzzball03 replied:
Yup, old comment. Left it for me to find while working on this.

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,79 @@
+  if (ok < 0) return HTTPD_CGI_DONE;
+
+  char *coll;
+  if (ok > 0) {
```
I heaven't dealt with the ajax stuff in a while and forgotten everything (oops), but I think you're gonna get one port at a time when you fill out one input field and hit enter.

fuzzball03 replied:
I think you're right. Ill modify to have it check the flashconfig variables.

tve reviewed (changes requested):
Looking promising! :-)

fuzzball03 committed:
Cgitelnet changes/fixes

Corrected several mistakes (16 vs 8bit int)
Check to see which variable we are modifying & fill the other one from
flag so we can make sure they do NOT conflict.
Added return error http header using existing collision goto.

fuzzball03 committed:
Cleanup

Remove unecessary comments

fuzzball03 committed:
Ajax/Html changes

Changes made so we can use a single namespace "/telnet" vs two.
Somewhat dirty fix inserted into existing ajax function.

fuzzball03 committed:
Cgitelnet update

Place variable declartion in correct place.

fuzzball03 committed:
Update cgitelnet.c

fuzzball03 committed:
Update cgitelnet.c

fuzzball03 committed:
Update cgitelnet.c

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
update for s3

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
Cgitelnet reform json output

fuzzball03 committed:
Update cgitelnet.c

fuzzball03 committed:
Update cgitelnet.c

fuzzball03 committed:
Update home.html

tve commented on .travis.yml:
```diff
@@ -1,31 +1,27 @@
-
+- curl -Ls http://s3.voneicken.com/xtensa-lx106-elf-20160330.tgx | tar Jxf -
+- curl -Ls http://s3.voneicken.com/esp_iot_sdk_v2.0.0.p1.tgx | tar -C .. -Jxf -
 after_script:
```
why all these formatting changes here? please don't make random formatting changes, it makes reviewing difficult, plus, I don't see why your formatting is better than mine...

tve commented on .travis.yml:
```diff
@@ -1,31 +1,27 @@
+  to $S3_BUCKET; curl -Ls https://github.com/rlmcpherson/s3gof3r/releases/download/v0.5.0/gof3r_0.5.0_linux_amd64.tar.gz
+  | tar zxf - gof3r_0.5.0_linux_amd64/gof3r; mv gof3r*/gof3r .; ls *.tgz | xargs -I
+  {} ./gof3r put -b $S3_BUCKET -k esp-link/{} --acl public-read -p {}; ls *.tgz |
+  xargs -I {} echo "URL: http://$S3_BUCKET/esp-link/{}"; fi'
```
what is this change about?

tve commented on .travis.yml:
```diff
@@ -1,31 +1,27 @@
+  file_glob: true
+  file: esp-link-master-*.t*gz
+  on:
+    repo: fuzzball03/esp-link
```
what is this change about?

fuzzball03 replied:
Ignore any Travis changes please.
I've done those changes so any commits made on my fork will compile via Travis without the need for me to build it myself.

tve replied:
mhh, I'm not a fan of having to go and clean-up after you ;-) but I also see that the uploads aren't working for pull requests. I had thought that they would. I think the solution is for me to provide my S3 keys to anyone who wants to have the builds done automatically.

tve commented on esp-link/config.c:
```diff
@@ -35,6 +35,8 @@ FlashConfig flashDefault = {
   .parity	= NONE_BITS,
   .stop_bits	= ONE_STOP_BIT,
+  .telnet_port1 = 23,
+  .telnet_port2 = 2323,
```
Something that seems to be missing is to take care of the upgrade path. When someone upgrades the telnet_portN fields in flash will be 0, so something needs to detect that when loading from flash and set the defaults explicitly.

fuzzball03 replied:
Ill take a look to see how that was done in some other code and proceed appropriately.

fuzzball03 replied:
@tve So I looked into this a bit - as the code stands now, if I erase the flash... Then flash my current build, the default ports of 23 & 2323 are set(as they are set in this file). How is it accomplishing this?

As far as I can tell, the configRestore() function is where checks for "0" values are checked. 
Why not simply use a loop to check all the flashConfig variables for "0" or NULL and fill them with their flashDefault values if true? (Just trying to learn here - I'm sure you or someone else has a good reason here)

tve commented on html/home.html:
```diff
@@ -50,6 +50,29 @@ <h2 id="version"></h2>
             <a href="https://github.com/jeelabs/esp-link/blob/master/README.md">the online README</a>
             for up-to-date help.</p>
           </div>
+<div class="card">
```
looks like the indent is off...

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,85 @@
+    }
+    
+    // apply the changes
+    serbridgeInit(flashConfig.telnet_port1, flashConfig.telnet_port2);
```
Something that is missing here is to call a new serbridgeStop function that closes the current listening sockets and resets any serbridge state so you can call serbridgeInit again. The latter wasn't designed to be called multiple times...
This may also explain some of the issues you have been seeing.

fuzzball03 replied:
That makes sense.
I did wonder how it was handling another call like that.
randy in #205 made some changes to the serbridgeinit function - ill look and see if perhaps he addressed this already.

fuzzball03 replied:
Citrection, #206 is the code I was referring to of Randys

tve commented on esp-link/cgitelnet.c:
```diff
@@ -0,0 +1,85 @@
+  uint16_t port1, port2;
+  ok |= getUInt16Arg(connData, "port1", &port1);
+  ok |= getUInt16Arg(connData, "port2", &port2);
+  if (ok < 0) { coll = "Failed to fetch ports. Are they valid?"; goto collision; }
```
This should perhaps be `<= 0`? As it is you're not returning any response when ok==0 and that's not good.
Second issue is that goto collision results in an rather poor error message.
Third issue is that having the string verbatim here ends up putting it into RAM, which is not plentiful. If you used sprintf the constant format string ends up in flash and doesn't consume RAM (there's some hacky magic that does that).

fuzzball03 replied:
That's certainly something I was unaware of. I will try a better approach based on your recommendations sir.

tve replied:
no problem. I should go through the whole code and see where I can push constant strings into flash. Can never have too much memory on this thing...

tve reviewed (changes requested):
Getting close :-)

fuzzball03 committed:
Update readme to reflect changes

fuzzball03 committed:
Various changes

Added new function to reset serial bridge.
Modified other serial bridge functions.
Add variables to later allow for the connection type to be
changed(secured, etc).
Set flash values for telnet to defaults if they are not set or =0.

fuzzball03 committed:
Fix a few errors..

fuzzball03 committed:
Serialbridge.c update

dannybackx committed:
Fix compile errors

dannybackx committed:
Fix compile errors

dannybackx committed:
Fix compile errors

dannybackx committed:
Fix compile errors

dannybackx committed:
Change comment

dannybackx committed:
Fix the build problem : don't call sprintf.

dannybackx committed:
Aha

dannybackx committed:
Change call to 0/1 instead of 1/2

fuzzball03 committed:
Makefile/Build file changes

fuzzball03 committed:
Formatting cleanup

fuzzball03 committed:
Modify all calls to port variables to match. Espconn1 = TelnetPort1

Should make code more fluid & inuitive.
All code changes made to reflect javascrupt, html, and C.

fuzzball03 committed:
Fix error for sucessful Travis build

fuzzball03 commented on Makefile:
```diff
@@ -52,12 +52,15 @@ ESP_HOSTNAME  ?= esp-link
-XTENSA_TOOLS_ROOT ?= $(abspath ../esp-open-sdk/xtensa-lx106-elf/bin)/
+XTENSA_TOOLS_ROOT ?= $(abspath ../espressif/xtensa-lx106-elf/bin)/
+$(warning Using XTENSA TOOLS from $(XTENSA_TOOLS_ROOT))
 
```
Please look at this commit for a more thorough explanation of changes made  in Makefile
https://github.com/fuzzball03/esp-link/commit/286e8865ebd60d17d050d948160095725a12f897

fuzzball03 committed:
Start making changes for Danny's security settings

fuzzball03 committed:
fix error

fuzzball03 committed:
Fix errors

fuzzball03 committed:
HTML Changes.
I may write a new javscript method all together for this... My goal might be to make one that is more adaptable to save on time & reduce overall size/complexity of code.

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
Update .travis.yml

fuzzball03 committed:
5

fuzzball03 committed:
9

fuzzball03 committed:
10

fuzzball03 committed:
12
